python/ray/spark/__init__.py

```diff
@@ -0,0 +1,593 @@
+    )
+    os.makedirs(ray_log_dir, exist_ok=True)
+
+    # TODO: Many ray processes logs are outputted under "{ray_temp_dir}/session_latest/logs",
+    #  Proposal: Update "ray start" scirpt to add a new option "ray_log_dir", and output logs
+    #  to a different directory specified by "ray_log_dir", instead of using
+    #  "{ray_temp_dir}/session_latest/logs".
+    #  The reason is, for ray on spark, user is hard to access log files on spark worker machines,
+    #  (especially on databricks runtime), so we'd better set the log output dir to be a
+    #  path mounted with NFS shared by all spark cluster nodes, so that the user can access
+    #  these remote log files from spark drive side easily.
+    ray_temp_dir = os.path.join(
+        ray_temp_root_dir, f"ray-temp-{ray_head_port}-{temp_dir_unique_suffix}"
+    )
+    os.makedirs(ray_temp_dir, exist_ok=True)
+
+    _logger.warning(
+        "Logs for ray head / worker starting script can be found in local disk path "
+        f"{ray_log_dir}. Logs for ray processes can be found in local disk path "
+        f"{ray_temp_dir}/session_latest/logs."
+    )
+
+    ray_head_node_cmd = [
+        ray_exec_path,
+        "start",
+        f"--temp-dir={ray_temp_dir}",
+        "--block",
+        "--head",
+        "--disable-usage-stats",
+        f"--port={ray_head_port}",
+        "--include-dashboard=false",
+        f"--num-cpus=0",  # disallow ray tasks scheduled to ray head node.
+        # limit the memory usage of head node because no task running on it.
+        f"--memory={128 * 1024 * 1024}",
+        # limit the object store memory usage of head node because no task running on it.
+        f"--object-store-memory={128 * 1024 * 1024}",
+        *_convert_ray_node_options(head_options),
+    ]
+
+    _logger.info(f"Start Ray head, command: {' '.join(ray_head_node_cmd)}")
+
+    with open(
+        os.path.join(ray_log_dir, "ray-start-head.log"), "w", buffering=1
+    ) as head_log_fp:
+        ray_head_proc = exec_cmd(
+            ray_head_node_cmd,
+            synchronous=False,
+            capture_output=False,
+            stream_output=False,
+            stdout=head_log_fp,
+            stderr=subprocess.STDOUT,
+        )
+
+    # wait ray head node spin up.
+    wait_ray_node_available(
+        ray_head_hostname,
+        ray_head_port,
+        40,
+        error_on_failure="Start Ray head node failed!",
+    )
+
+    _logger.info("Ray head node started.")
+
+    # NB:
+    # In order to start ray worker nodes on spark cluster worker machines,
+    # We launch a background spark job:
+    #  1. it is a barrier mode spark job, i.e. all spark tasks in the job run concurrently.
+    #     If the spark cluster resources are not sufficient to launch all tasks concurrently,
+    #     the spark job will hang and retry. If spark exceeds the maximum retry count, the
+    #     submitted operation will fail.
```

Comment:
Comment needs update, now we use normal spark job and the behavior changed.

Reply:
adjusted